overlord/devicestate/handlers_install.go

```diff
@@ -1252,45 +1298,37 @@ func (m *DeviceManager) doInstallFinish(t *state.Task, _ *tomb.Tomb) error {
 	if err := t.Get("on-volumes", &onVolumes); err != nil {
 		return err
 	}
-	logger.Debugf("starting install-finish for %q systemd on %v", systemLabel, onVolumes)
 
-	essentialTypes := []snap.Type{snap.TypeKernel, snap.TypeBase, snap.TypeGadget}
+	var encryptSetupData *install.EncryptionSetupData
+	cached := st.Cached(encryptionSetupDataKey{systemLabel})
+	if cached != nil {
+		var ok bool
+		// TODO check that encryptSetupData is not out of sync with the onVolumes we get
+		encryptSetupData, ok = cached.(*install.EncryptionSetupData)
+		if !ok {
+			return fmt.Errorf("internal error: wrong data type under encryptionSetupDataKey")
+		}
+	}
+
 	st.Unlock()
-	sys, snapInfos, snapSeeds, err := m.loadSystemAndEssentialSnaps(systemLabel, essentialTypes)
+	sys, snapInfos, snapSeeds, mntPtForType, unmount, err := m.loadAndMountSystemLabelSnaps(systemLabel)
 	st.Lock()
 	if err != nil {
 		return err
 	}
-	// Unset revision here actually means that the snap is local.
-	// Assign then a local revision as seeding/installing the snap would do.
-	for _, typ := range essentialTypes {
-		snInfo := snapInfos[typ]
-		if snInfo.Revision.Unset() {
-			snInfo.Revision = snap.R(-1)
-		}
-	}
+	defer unmount()
 
-	// Mount gadget and kernel
-	mntPtForType := make(map[snap.Type]string)
-	for _, seedSn := range []*seed.Snap{snapSeeds[snap.TypeGadget], snapSeeds[snap.TypeKernel]} {
-		mntPt, unmountSnap, err := mountSeedSnap(seedSn)
-		if err != nil {
-			return err
-		}
-		mntPtForType[seedSn.EssentialType] = mntPt
-		defer func() {
-			errRest := unmountSnap()
-			if err == nil {
-				err = errRest
-			}
-		}()
+	// TODO validation of onVolumes versus gadget.yaml
+
+	// Check if encryption is mandatory
+	if sys.Model.Grade() != asserts.ModelDangerous && encryptSetupData == nil {
```

Comment:
this doesn't look like the right check, we probably need to call encryptionSupportInfo again or organize some subset of it